train.py

```diff
@@ -72,10 +85,69 @@ def loss_for_permutation(
             loss += best_loss
     return loss, best_mask_order
 
+
+def save_ckpt(ckpt_file: str, model: nn.Module,
+              optimizer: Optional[th.optim.Optimizer] = None):
+    ckpt_file = Path(ckpt_file)
+    _ensure_dir(ckpt_file.parent)
+    save_dict: Dict[str, Any] = {}
+    save_dict['model'] = model.state_dict()
+    if optimizer is not None:
+        save_dict['optimizer'] = optimizer.state_dict()
+    th.save(save_dict, str(ckpt_file))
+
+
+def load_ckpt(ckpt_file: str, model: nn.Module,
+              optimizer: Optional[th.optim.Optimizer] = None):
+    ckpt_file = Path(ckpt_file)
+    save_dict = th.load(str(ckpt_file))
+    model.load_state_dict(save_dict['model'])
+    if optimizer is not None:
+        optimizer.load_state_dict(save_dict['model'])
+
+
+def add_tensorboard_graph(model: nn.Module, loader: DataLoader,
+                          writer: SummaryWriter,
+                          device: th.device):
+    try:
+        train_mode = model.training
+        model.train(False)
+        # data = dataset[0][0]
+        data = iter(loader).next()[0]
+        #data = {k: th.as_tensor(v).to(device)[None]
+        #        for k, v in data.items()}
+        #data['prv_state'] = th.zeros(
+        #    size=[1, 8, 128, 128, 48],
+        #    dtype=th.float32,
+        #    device=data['tsdf'].device)
+
+        keys = sorted(list(data.keys()))
+
+        class DummyModel(nn.Module):
+            def __init__(self, model: nn.Module):
+                super().__init__()
+                self.model = model
+
+            def forward(self, *args):
+                return tuple(self.model(
+                    {k: v.to(device) for k, v in zip(keys, args)}).values())
+        writer.add_graph(DummyModel(model),
+                         tuple(data[k].detach() for k in keys))
+        writer.flush()
+        # In case this op leaves any artifacts...
+        th.cuda.empty_cache()
+    finally:
+        model.train(train_mode)
+
+
 def main():
     parser = ArgumentParser()
     parser.add_arguments(Config, dest='cfg')
     cfg = parser.parse_args().cfg
+    root = Path(cfg.path)
+    index = len([d for d in root.glob('run-*') if d.is_dir()])
+    path = _ensure_dir(Path(cfg.path) / F'run-{index:03d}')
```

Comment:
I think it would be good to print the generated subdirectory.

Reply:
I'll do that after the meeting.